Mathlib/NumberTheory/Padics/Complex.lean

```diff
@@ -0,0 +1,194 @@
+
+/-- `PadicAlg p` is a valued field, with the valuation corresponding to the `p`-adic norm. -/
+instance valuedField : Valued (PadicAlg p) ℝ≥0 := NormedField.toValued
+
+/-- The valuation of `x : PadicAlg p` agrees with its `ℝ≥0`-valued norm. -/
+theorem valuation_def (x : PadicAlg p) : Valued.v x = ‖x‖₊ := rfl
+
+/-- The coercion of the valuation of `x : PadicAlg p` to `ℝ` agrees with its norm. -/
+theorem valuation_coe (x : PadicAlg p) :
+  ((Valued.v x : ℝ≥0) : ℝ) = spectralNorm ℚ_[p] (PadicAlg p) x := rfl
+
+/-- The valuation of `p : PadicAlg p` is `1/p`. -/
+theorem valuation_p (p : ℕ) [Fact p.Prime] : Valued.v (p : PadicAlg p) = 1 / (p : ℝ≥0) := by
+  rw [← map_natCast (algebraMap ℚ_[p] (PadicAlg p))]
+  ext
+  rw [valuation_coe, spectralNorm_extends, padicNormE.norm_p, one_div, NNReal.coe_inv,
+    NNReal.coe_natCast]
+
+end PadicAlg
+
+/-- `ℂ_[p]` is the field of `p`-adic complex numbers, that is, the completion of `PadicAlg p` with
+respect to the `p`-adic norm. -/
+def PadicComplex := UniformSpace.Completion (PadicAlg p)
+
+/-- `ℂ_[p]` is the field of `p`-adic complex numbers. -/
+notation "ℂ_[" p "]" => PadicComplex p
+
+namespace PadicComplex
+
+/-- The `p`-adic complex numbers have a field structure. -/
+instance : Field ℂ_[p] := UniformSpace.Completion.instField
+
+/-- The `p`-adic complex numbers are inhabited. -/
+instance : Inhabited ℂ_[p] := ⟨0⟩
+
+/-- `ℂ_[p]` is a valued field, where the valuation is the one extending that on `PadicAlg p`. -/
+instance valued : Valued ℂ_[p] ℝ≥0 := Valued.valuedCompletion
+
+/-- `ℂ_[p]` is a complete space. -/
+instance completeSpace : CompleteSpace ℂ_[p] := UniformSpace.Completion.completeSpace _
+
+instance : Coe (PadicAlg p) ℂ_[p] := ⟨UniformSpace.Completion.coe'⟩
+
+/-- The valuation on `ℂ_[p]` extends the valuation on `PadicAlg p`. -/
+theorem valuation_extends (x : PadicAlg p) : Valued.v (x : ℂ_[p]) = Valued.v x :=
+  Valued.extension_extends _
+
+/-- `ℂ_[p]` is an algebra over `PadicAlg p`. -/
+instance : Algebra (PadicAlg p) ℂ_[p] := UniformSpace.Completion.algebra' _
+
+theorem coe_eq (x : PadicAlg p) : (x : ℂ_[p]) = algebraMap (PadicAlg p) ℂ_[p] x := rfl
+
+theorem coe_zero : ((0 : PadicAlg p) : ℂ_[p]) = 0 := rfl
+
+/-- `ℂ_[p]` is an algebra over `ℚ_[p]`. -/
+instance : Algebra ℚ_[p] ℂ_[p] :=
+  ((algebraMap (PadicAlg p) ℂ_[p]).comp (algebraMap ℚ_[p] (PadicAlg p))).toAlgebra
+
+instance : IsScalarTower ℚ_[p] (PadicAlg p) ℂ_[p] := IsScalarTower.of_algebraMap_eq (congrFun rfl)
+
+@[simp, norm_cast]
+lemma coe_natCast (n : ℕ) : ((n : PadicAlg p) : ℂ_[p]) = (n : ℂ_[p])  := by
+  rw [← map_natCast (algebraMap (PadicAlg p) ℂ_[p]) n, coe_eq]
+
+/-- The valuation of `p : ℂ_[p]` is `1/p`. -/
+theorem valuation_p : Valued.v (p : ℂ_[p]) = 1 / (p : ℝ≥0) := by
+  have := valuation_extends p (p : PadicAlg p)
+  rw [PadicAlg.valuation_p] at this
+  simpa only [← map_natCast (algebraMap (PadicAlg p) ℂ_[p]), ← coe_eq]
```

Comment:
```suggestion
  simpa [← map_natCast (algebraMap (PadicAlg p) ℂ_[p]), ← coe_eq]
```

Reply:
I cannot remove this `only`, that leads to a "maximum recursion depth has been reached" error.

Reply:
I have golfed the proof so that the `simpa` is no longer needed.